torch/nn/_functions/vision.py

```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
     @staticmethod
     def forward(ctx, theta, size):
         assert type(size) == torch.Size
-        N, C, H, W = size
-        ctx.size = size
-        if theta.is_cuda:
-            AffineGridGenerator._enforce_cudnn(theta)
-            assert False
-        ctx.is_cuda = False
-        base_grid = theta.new(N, H, W, 3)
-        linear_points = torch.linspace(-1, 1, W) if W > 1 else torch.Tensor([-1])
-        base_grid[:, :, :, 0] = torch.ger(torch.ones(H), linear_points).expand_as(base_grid[:, :, :, 0])
-        linear_points = torch.linspace(-1, 1, H) if H > 1 else torch.Tensor([-1])
-        base_grid[:, :, :, 1] = torch.ger(linear_points, torch.ones(W)).expand_as(base_grid[:, :, :, 1])
-        base_grid[:, :, :, 2] = 1
-        ctx.base_grid = base_grid
-        grid = torch.bmm(base_grid.view(N, H * W, 3), theta.transpose(1, 2))
-        grid = grid.view(N, H, W, 2)
+
+        if len(size) == 5:
+            N, C, D, H, W = size
+            ctx.size = size
+            ctx.is_cuda = theta.is_cuda
+            base_grid = theta.new(N, D, H, W, 4)
+
+            w_points = (torch.linspace(-1, 1, W) if W > 1 else torch.Tensor([-1]))
+            h_points = (torch.linspace(-1, 1, H) if H > 1 else torch.Tensor([-1])).unsqueeze(-1)
+            d_points = (torch.linspace(-1, 1, D) if D > 1 else torch.Tensor([-1])).unsqueeze(-1).unsqueeze(-1)
+
+            base_grid[:, :, :, :, 0] = w_points
+            base_grid[:, :, :, :, 1] = h_points
+            base_grid[:, :, :, :, 2] = d_points
+            base_grid[:, :, :, :, 3] = 1
+            ctx.base_grid = base_grid
+            grid = torch.bmm(base_grid.view(N, D * H * W, 4), theta.transpose(1, 2))
+            grid = grid.view(N, D, H, W, 3)
+
+        elif len(size) == 4:
+            N, C, H, W = size
+            ctx.size = size
+            if theta.is_cuda:
+                AffineGridGenerator._enforce_cudnn(theta)
+                assert False
+            ctx.is_cuda = False
```

Comment:
This `if theta.is_cuda:` check seems needlessly restrictive. The `def affine_grid_generator(theta, size):` function handles the dispatch between CPU and CUDA; why repeat the check here as well?